### PR TITLE
Add ranges::rotate_copy

### PR DESCRIPTION
**Pull request by akukanov**
`ranges::rotate_copy` is implemented according to the [oneDPL specification](https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneDPL/source/parallel_api/parallel_range_api.html#sequence-reordering). The return value semantics matches the [accepted C++26 design](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p3709r0.html#rotate_copy_design).

The implementation reuses the existing iterator-based patterns, extending them with an extra parameter for the output size; `dpl::rotate_copy` uses the maximal possible size_t value for it. The parallel implementation for CPU (see `algorithm_imp.h`) is changed to iterate over the ouput rather than the input (with the effective output size first adjusted to the minimum of two sizes); that ensures no special checks for out-of-bound writes. The SYCL implementation reworks `__rotate_copy` kernel function (see `unseq_backend_sycl.h`) to separate the roles of `__size` as the iteration space limit (set to the minimum of two range sizes) and `__size_in` as the input range size necessary to compute the wraparound point.

The test uses a special wrapper over the algorithm call in order to adjust the parameter order for the existing harness in `std_ranges_test.h`, and also to set the desired pivot point without knowing the actual range sizes (which are set by the harness). The checker is custom-written as the existing standard algorithms do not provide the necessary semantics, but the implementation is rather straightforward (and very similar to the possible implementation provided in the referenced C++26 design paper) and in my opinion does not need a special self-check.

### CODE REVIEW

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h
Comment (rarutyun) [inline include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:23]: `#include <cassert>` is missing

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 1]
```diff
@@ -1919,68 +1919,74 @@ __pattern_rotate(__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec, _RandomAc
...
-__brick_rotate_copy(_Tag, _ExecutionPolicy&&, _ForwardIterator __first, _ForwardIterator __middle,
-                    _ForwardIterator __last, _OutputIterator __result) noexcept
+__brick_rotate_copy(_ForwardIterator __first, _ForwardIterator __middle, _ForwardIterator __last,
+                    _OutputIterator __result, std::size_t __n_out, /*is_vector=*/std::false_type) noexcept
 {
-    static_assert(__is_serial_tag_v<_Tag> || __is_parallel_forward_tag_v<_Tag>);
-
-    return ::std::rotate_copy(__first, __middle, __last, __result);
+    std::size_t __i = 0;
+    for (_ForwardIterator __from = __middle; __from != __last && __i < __n_out; ++__i)
```
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/algorithm_impl.h:1928]: @akukanov, please remind me, why we can't use here two calls of `std::copy` with proper `last` state? Because we will move iterator till the end for `last` evaluation if the our iterators aren't random access?

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h [hunk 2]
```diff
@@ -1197,6 +1197,40 @@ struct __internal::__reverse_copy_fn
...
+    std::ranges::in_in_out_result<std::ranges::borrowed_iterator_t<_InRange>,
+                                  std::ranges::borrowed_iterator_t<_InRange>,
+                                  std::ranges::borrowed_iterator_t<_OutRange>>
+    operator()(_ExecutionPolicy&& __exec, _InRange&& __in_r, std::ranges::iterator_t<_InRange> __middle,
+               _OutRange&& __out_r) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
+        auto [__first_in, __last_in, __in_size] = oneapi::dpl::__ranges::__bounds_and_size(__in_r);
+        auto __first_out = std::ranges::begin(__out_r);
+        const std::size_t __min_size = std::min<std::size_t>(__in_size, std::ranges::size(__out_r));
```
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:1220]:
```suggestion
        auto [__first_out, __min_size] = oneapi::dpl::__ranges::__begin_and_size(__out_r);
```

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 3]
```diff
@@ -1919,68 +1919,74 @@ __pattern_rotate(__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec, _RandomAc
...
     using __backend_tag = typename __parallel_tag<_IsVector>::__backend_tag;
 
+    if (__n_out > std::size_t(__last - __first))
+        __n_out = __last - __first;
+
     return __internal::__except_handler([&]() {
         __par_backend::__parallel_for(
-            __backend_tag{}, ::std::forward<_ExecutionPolicy>(__exec), __first, __last,
-            [__first, __last, __middle, __result](_RandomAccessIterator1 __b, _RandomAccessIterator1 __e) {
+            __backend_tag{}, std::forward<_ExecutionPolicy>(__exec), std::size_t(0), __n_out,
```
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/algorithm_impl.h:1973]:
```suggestion
            __backend_tag{}, std::forward<_ExecutionPolicy>(__exec), std::size_t{0}, __n_out,
```

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 4]
```diff
@@ -1919,68 +1919,74 @@ __pattern_rotate(__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec, _RandomAc
...
 _RandomAccessIterator2
 __pattern_rotate_copy(__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec, _RandomAccessIterator1 __first,
-                      _RandomAccessIterator1 __middle, _RandomAccessIterator1 __last, _RandomAccessIterator2 __result)
+                      _RandomAccessIterator1 __middle, _RandomAccessIterator1 __last, _RandomAccessIterator2 __result,
+                      std::size_t __n_out)
 {
     using __backend_tag = typename __parallel_tag<_IsVector>::__backend_tag;
 
+    if (__n_out > std::size_t(__last - __first))
+        __n_out = __last - __first;
```
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/algorithm_impl.h]:
```suggestion
    __n_out = std::min(__n_out, std::size_t(__last - __first));
```

File: test/parallel_api/ranges/std_ranges_rotate_copy.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                                                       std::ranges::borrowed_iterator_t<decltype(r_out)>>;
+        return ret_type{it1, it2, it_out};
+    };
+
+    const int test_sz = 19787;
+    test_range_algo<0, int, data_in_out_lim>{big_sz}(rotate_copy_tester, rotate_copy_checker);
+    test_range_algo<1, float, data_in_out_lim>{test_sz}(rotate_copy_tester, rotate_copy_checker, test_sz - 1);
+    test_range_algo<2, P2, data_in_out_lim>{test_sz/11}(rotate_copy_tester, rotate_copy_checker, 0);
+    test_range_algo<3, int, data_in_out_lim>{test_sz/17}(rotate_copy_tester, rotate_copy_checker, 1);
+#endif //_ENABLE_STD_RANGES_TESTING
```
Comment (SergeyKopienko) [inline test/parallel_api/ranges/std_ranges_rotate_copy.pass.cpp]: AI comment:
- In the ranges test (std_ranges_rotate_copy.pass.cpp), I don't see an explicit dedicated case where `n_out` lands exactly on the `__split` boundary (i.e., where the truncation happens precisely at the junction between the first and second chunk). The combination of `pivot_pos` and `n_out = r_size/2` might incidentally produce this, but there's no targeted test specifically verifying `n_out == last-middle` exactly. This is an edge case in the code (`__e >= __split`, `__b < __split`) where both branches can fire on the same task chunk simultaneously — it would be worth checking explicitly.

File: include/oneapi/dpl/pstl/algorithm_impl.h (hunk 1 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/algorithm_impl.h:1928]: Yes, we cannot pre-compute the stop position for forward iterators.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/algorithm_impl.h:1928]: ok

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h (hunk 2 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:1220]: You'd lose `std::min` this way and would have to apply it separately.

File: include/oneapi/dpl/pstl/algorithm_impl.h (hunk 4 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/algorithm_impl.h]: Changed. Though it's always easier for me personally to understand the "cap a variable by some value" code with "if + assign" rather than with `std::min`.

File: test/parallel_api/ranges/std_ranges_rotate_copy.pass.cpp (hunk 5 above)
Comment (akukanov) [reply test/parallel_api/ranges/std_ranges_rotate_copy.pass.cpp]: Added a test with the pivot set to `test_sz/2`

**Review by SergeyKopienko — DISMISSED** [review]
LGTM

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h [hunk 6]
```diff
@@ -1679,31 +1679,32 @@ __pattern_rotate(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Iterator
...
-    auto __n = __last - __first;
-    if (__n <= 0)
+template <typename _BackendTag, typename _ExecutionPolicy, typename _RandomAccessIterator1,
+          typename _RandomAccessIterator2>
+_RandomAccessIterator2
+__pattern_rotate_copy(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _RandomAccessIterator1 __first,
+                      _RandomAccessIterator1 __new_first, _RandomAccessIterator1 __last,
+                      _RandomAccessIterator2 __result, std::size_t __n_out)
+{
+    std::size_t __n = __last - __first;
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]:
```suggestion
    const std::size_t __n = __last - __first;
```
Otherwise, it's inconsistent with the chosen approach

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:23]: Fixed.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h (hunk 6 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]: Applied.

**Review by SergeyKopienko — APPROVED** [review]
LGTM
